### ISSUE TITLE
Search with a dash fails

### ISSUE THREAD

**Issue by vhpoet** [typesense/typesense#287]
## Description

I have a document that typesense doesn't find on search. I've tried both `4 hour` and `4-hour` queries, both don't bring up the book (`workweek` works though). It works if I do `"drop_tokens_threshold": 1` but that's not desirable.

**Document**
```js
{
    "id": "7cbf8757-9046-4f0e-bb93-dd8e6fdda183",
    "title": "The 4-Hour Workweek",
    "slug": "the-4-hour-workweek",
    "weight": 7929
}
```
**Search**
```js
{
    "collection": "books",
    "q": "4-hour",
    "query_by": "title",
    "num_typos": 1,
    "drop_tokens_threshold": 0
}
```
**Response**
```js
{
    "facet_counts": [],
    "found": 0,
    "hits": [],
    "out_of": 1,
    "page": 1,
    "request_params": {
        "collection_name": "books",
        "per_page": 10,
        "q": "4-hour"
    },
    "search_time_ms": 0
}
```

## Metadata

**Typsense Version**: `0.20.0` and `0.21.0.rc10`

**OS**: `osx` and `centos`

**Comment by kishorenc** [comment]
@vhpoet This does work for me, [see here](https://gist.github.com/kishorenc/f86eda7beb5a2d4e57ab0703af258f11).

Can you please produce a similar snippet which show cases the issue?

**Comment by pyr0hu** [comment]
@kishorenc the example you provided works as intended, but if you try to use `num_typos=0`, then it won't find the document, although there is no typo in it

`curl -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" http://localhost:8108/collections/books/documents/search/\?q\=4-hour\&query_by\=title\&num_typos\=0`

**Comment by vhpoet** [comment]
@kishorenc have you tried with `drop_tokens_threshold=0`? I think that's causing it. So your exact snippet, but for the search you do

```
curl -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" http://localhost:8108/collections/books/documents/search/\?q\=4-hour\&query_by\=title\&drop_tokens_threshold\=0
```

**Comment by kishorenc** [comment]
@vhpoet There is a regression in 0.20 with queries containing hyphens (as a side effect of adding `-word` negation support). I've fixed that in this RC build: `typesense/typesense:0.21.0.rc15`

If you can confirm that it is fixed, I can build a RPM for you again.

**Comment by vhpoet** [comment]
@kishorenc that works, I'd appreciate an RPM build.

**Comment by kishorenc** [comment]
@vhpoet Cool, here's the latest RC [RPM build](https://dl.typesense.org/releases/0.21.0.rc18/typesense-server-0.21.0.rc18-1.x86_64.rpm).

**Comment by kishorenc** [comment]
@vhpoet Noticed another minor issue, which I have fixed, so please use [`0.21.0.rc19`](https://dl.typesense.org/releases/0.21.0.rc19/typesense-server-0.21.0.rc19-1.x86_64.rpm). I've also updated my previous comment above with the correct link.

**Comment by vhpoet** [comment]
@kishorenc I've updated to rc19, `4-hour` works now, but `4 hour` doesn't. Is that intended?

**Comment by kishorenc** [comment]
@vhpoet Yes, `4hour` cannot match `4 hour` unless `drop_tokens_threshold` is set to `1`. Otherwise, Typesense is unable to drop the tokens to query individual tokens for typo match. With `drop_tokens_threshold: 1` the record will be surfaced.

**Comment by kishorenc** [comment]
Shipped in `0.21.0` release.
